scrapy/linkextractors/htmlparser.py

```diff
@@ -39,7 +40,7 @@ def _extract_links(self, response_text, response_url, response_encoding):
         ret = []
         base_url = urljoin(response_url, self.base_url) if self.base_url else response_url
         for link in links:
-            if isinstance(link.url, unicode):
+            if isinstance(link.url, six.text_type):
```

Comment:
six.text_type is unicode string; "native" means "str" which means bytes in 2.x and unicode in 3.x.

The problem is not with this check per se, the problem is that link.url is encoded to bytes after this check. I think it should be converted to str using `to_native_str()`. But to do it we need to check all Link usages and docs - it should be consistent.